packages/dds/merge-tree/src/test/client.applyStashedOpFarm.spec.ts

```diff
@@ -0,0 +1,192 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import random from "random-js";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IMergeTreeOp } from "../ops";
+import { SegmentGroup } from "../mergeTree";
+import {
+    generateClientNames,
+    doOverRange,
+    runMergeTreeOperationRunner,
+    annotateRange,
+    removeRange,
+    IMergeTreeOperationRunnerConfig,
+    IConfigRange,
+} from "./mergeTreeOperationRunner";
+import { TestClient } from "./testClient";
+import { TestClientLogger } from "./testClientLogger";
+
+// This test is based on reconnectFarm, but we keep a second set of clients. For
+// these clients, we apply the generated ops as stashed ops, then regenerate
+// them to simulate resubmit(), then apply them. In the end, they should arrive
+// at the same state as the "normal" set of clients
+let stashClients: TestClient[] = [];
+
+function applyMessagesWithReconnect(
+    startingSeq: number,
+    messageDatas: [ISequencedDocumentMessage, SegmentGroup | SegmentGroup[], number][],
+    clients: readonly TestClient[],
+    logger: TestClientLogger,
+) {
+    let seq = startingSeq;
+    const reconnectClientMsgs: [IMergeTreeOp, SegmentGroup | SegmentGroup[]][] = [];
+    let minSeq = 0;
+
+    // apply ops as stashed ops except for client #1
+    const stashedOps = [];
+    for (const messageData of messageDatas) {
+        if (messageData[2] !== 1) {
```

Comment:
rather than adding another element to the messageDatas array could you just check the client id on the sequenced message. so here
```suggestion
        if (messageData[0].clientId !== clients[1].longClientId) {
```

Reply:
i prefer this as it's one less piece of data to track in an already complex set of tests

Reply:
Below I also use it to index into the stash client array, which would not be super efficient if I had to loop over the array each time

Reply:
actually if there are at most 8 clients it's probably fine

Reply:
you could also build the map once, and keep it, rather than re-iterating the client of client every time, but either way is fine, i don't ever expect millions of clients so should always be fast enough.